### ISSUE TITLE
Performance of the tokenizer and V8 concatenated string

### ISSUE THREAD

**Issue by pmalouin** [harttle/liquidjs#161]
In our project, we are using the `engine.parse()` method to prepare templates and keep them in memory until they are needed. In our scenario, the same template is usually rendered many times so it makes sense to keep them in memory to get a better render speed.

When running some V8 heap snapshots of our server, we were seeing quite a large retained size from these template object. Of course, templates can be big, but there is an exponential multiplying factor as templates grow bigger.

When digging into the snapshot, it seems that for a given template, we retain multiple copies of the raw string:

![image](https://user-images.githubusercontent.com/1411117/66160309-dda87b80-e5f7-11e9-8cca-dca7d064df7b.png)

Or rather, we are retaining concatenated strings of the whole template:
![image](https://user-images.githubusercontent.com/1411117/66160374-0597df00-e5f8-11e9-9fc1-c2bdffcdbde6.png)

The explanation that I found is that the [Tokenizer](https://github.com/harttle/liquidjs/blob/master/src/parser/tokenizer.ts) builds the token strings by consuming the input string one character at a time. As a result, the underlying memory representation of the final string is a tree of "Concatenated strings" like we see in the memory snapshots.

The memory overhead of this memory representation is costly and I started looking at ways to "flatten" the string. I came about these stackoverflow discussions: https://stackoverflow.com/questions/40637838/how-to-flatten-v8-consstring and https://stackoverflow.com/questions/35354801/why-does-v8-run-out-of-memory-in-this-situation. The solutions are usually a bit hackish or fragile because they can break when changing the v8 version.

I've played around with some approach to trigger the string flattening in this branch: https://github.com/pmalouin/liquidjs/commit/42319215b6dbd31283d110249064cc4b49890ad4
That branch includes a few elements:
- I altered the demo script to show how we can measure the heap size and generate heap snapshots to dig into the memory representation.
- You can see that I added a `Number(str)` expression in the `Token` constructor: https://github.com/pmalouin/liquidjs/commit/42319215b6dbd31283d110249064cc4b49890ad4#diff-d9028715a3d61180b038a6f3f5685210R13

Adding this expression had a significant impact on the retained heap size. In a test scenario that is shown in that commit, heap usage went from `~14.9MB` to `~2.5MB`. I tested on Node 8 and Node 10 with similar results. I also ran the benchmarks and there seems to be a slight improvement to throughput, but maybe not statistically significant.

### Potential solutions

Triggering the flattening mechanism seems to be hard because it depends on the internals of v8. As the internals of v8 change, it's possible that an optimization will need to be adjusted. You can see from the commit history of [this project](https://github.com/davidmarkclements/flatstr) that this is a moving target. One option could be to implement this approach and maintain it.

Another approach could be to alter how the Tokenizer works. There might be a way to change how the token strings are built so that they are less prone to result in the "Concatenated String" memory representation. I haven't looked at this approach in details.

Because the measured impact of this issue is significant, do you believe we should look at this a bit further?

**Comment by harttle** [comment]
Another approach is using Buffer instead of string concat, but that will require more polyfills when liquidjs is used in browersers. A better solution is to maintain a browser version (with string concat) and a nodejs version (with Buffer) just like the ifs.ts, which is non-trivial and seems not worth it to be so complicated.

If your approach does the work, feel free to file a PR. Make sure your commit message aligns to https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits to allow semantic-release bot do the work.

**Comment by pmalouin** [comment]
I tested the `Number(str)` approach a bit further, and it seems that it does not trigger flattening under Node 12.x. I also tried `str | 0` which is the current implementation in `flatstr`, but this expression does not work in any version of Node that I tested.

I have the impression that going with Buffers will be the best solution for this. A quick test with `Buffer.from(str).toString()` seems to work reliably, but I haven't verified the throughput impact yet.
